### PR TITLE
Fix/windows ls native

### PR DESCRIPTION
**Pull request by seansie0830**
# Feat: Native Windows `ls` Fallback with SRP-based Refactor

## Summary
This PR implements a native Rust fallback for the `ls` command on Windows environments and refactors the underlying logic to adhere to the Single Responsibility Principle (SRP).

### 1. Windows Native Support
Windows lacks a native `ls` binary. To ensure `rtk` remains functional across all environments without external dependencies, this PR introduces a native implementation leveraging `std::fs`. This ensures the core directory listing functionality is available out-of-the-box on Windows.

### 2. SRP Refactor & Modular Architecture
The implementation has been decoupled into a dedicated `ls_win` module to improve maintainability and testability. The logic is now partitioned into discrete units:
- **`fetch_entries`**: Responsible for filesystem traversal and metadata retrieval.
- **`synthesize_output`**: A pure transformation layer that handles sorting, extension aggregation, and output formatting.
- **`run_native`**: The entry-point orchestrator managing timing and terminal I/O.
- **`LsRecord` Struct**: Serves as the data contract between the filesystem and the synthesizer.

### 3. Platform Consistency & Fixes
- **Filter Alignment**: Standardized the Windows filtering logic to match Unix behavior. Generic dot-file hiding has been replaced with targeted `NOISE_DIRS` filtering, ensuring high-signal files (like `.cargo-lock`) remain visible while maintaining LLM-friendly output.
- **Summary Synthesis Fix**: Corrected a formatting regression in the extension summary where the closing parenthesis was prematurely appended when truncating the extension list.

## Verification
- [ ] Validated native `ls` execution on Windows (fallback triggers when binary is missing).
- [ ] Verified output consistency between Windows and Linux platforms (filtering and formatting).
### test log  

for windows platform
```
> .\target\release\rtk.exe ls
.claude/
.github/
.rtk/
Formula/
docs/
hooks/
openclaw/
scripts/
src/
tests/
.gitignore  553B
.release-please-manifest.json  23B
CHANGELOG.md  74.2K
CLAUDE.md  7.4K
CONTRIBUTING.md  12.6K
Cargo.lock  48.8K
Cargo.toml  1.7K
DISCLAIMER.md  2.2K
INSTALL.md  11.4K
LICENSE  10.7K
README.md  20.5K
README_es.md  5.4K
README_fr.md  6.7K
README_ja.md  5.8K
README_ko.md  5.4K
README_zh.md  5.3K
SECURITY.md  7.6K
build.rs  2.4K
install.sh  2.9K
release-please-config.json  191B

Summary: 20 files, 10 dirs (12 .md, 2 .json, 1 .lock, 1 no ext, 1 .sh, +3 more)

```
for ubuntu platform
```
$ ./target/release/rtk ls
.claude/
.github/
.rtk/
Formula/
docs/
hooks/
openclaw/
scripts/
src/
tests/
.gitignore  553B
.release-please-manifest.json  23B
CHANGELOG.md  74.2K
CLAUDE.md  7.4K
CONTRIBUTING.md  12.6K
Cargo.lock  48.8K
Cargo.toml  1.7K
DISCLAIMER.md  2.2K
INSTALL.md  11.4K
LICENSE  10.7K
README.md  20.5K
README_es.md  5.4K
README_fr.md  6.7K
README_ja.md  5.8K
README_ko.md  5.4K
README_zh.md  5.3K
SECURITY.md  7.6K
build.rs  2.4K
install.sh  2.9K
release-please-config.json  191B

Summary: 20 files, 10 dirs (12 .md, 2 .json, 1 .gitignore, 1 .toml, 1 .lock, +3 more)
```

### CODE REVIEW

**Comment by pszymkowiak** [comment]
> [w] **wshm** · Automated triage by AI

## 📊 Automated PR Analysis

| | |
|---|---|
| ✨ **Type** | `feature` |
| 🟢 **Risk** | `low` |

### Summary

Adds a native Rust fallback for the `ls` command on Windows when no `ls` binary is available, using std::fs for directory listing. The implementation is split into a new `ls_win` module with separate functions for filesystem traversal, output synthesis, and orchestration following SRP principles.

### Review Checklist
- [ ] Tests present
- [ ] Breaking change
- [ ] Docs updated

---
*Analyzed automatically by [wshm](https://github.com/wshm-dev/wshm)* · This is an automated analysis, not a human review.

**Review by F0rty-Tw0 — COMMENTED** [review]
Solves a real problem: Windows users without Git Bash or WSL hit "ls not found" today, so the native fallback is the right shape of fix. The module structure (`fetch_entries` / `synthesize_output` / `run_native`) is a reasonable separation, and `LsRecord` is a good data contract.

Five inline findings follow. Two P0 blockers (no tests, parallel formatter that should be unified), two P1s on flag-drop and runner-bypass, and one P2 on output divergence visible in the PR's own test logs.

Three-OS CI green is encouraging, but the new module has no tests, so green here means existing tests still pass rather than new code is verified.

File: src/cmds/system/ls_win.rs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+                size: metadata.len(),
+                name,
+            });
+        }
+    }
+    Ok(records)
+}
+
+/// Synthesizes the compact, token-optimized string from a list of records.
+pub fn synthesize_output(records: Vec<LsRecord>) -> (String, String) {
```
Comment (F0rty-Tw0) [inline src/cmds/system/ls_win.rs]: Per `.claude/rules/cli-testing.md`, all new filters require snapshot and token-savings tests. The existing `ls.rs` has 17 unit tests covering parsing edge cases (symlinks, tied groups, year-format dates, filenames with spaces). This new module adds 157 lines with zero tests, including the entire output-synthesis path that drives the user-visible Windows experience.

Concrete asks:

- **Snapshot test for `synthesize_output`** with a hand-built `Vec<LsRecord>` fixture covering the dirs-first / files-with-size / summary-with-extensions output shape (mirror `test_compact_basic` and `test_compact_summary` in ls.rs).
- **`fetch_entries` test** using `tempfile::tempdir()` to create a tree with NOISE_DIRS, dotfiles, nested dirs, and verify the right entries get filtered and classified.
- **Token-savings assertion** comparing `synthesize_output` output vs a verbose `ls -la`-equivalent baseline string, asserting `>= 60%` savings per the project standard.

File: src/cmds/system/ls_win.rs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+fn get_extension(name: &str) -> String {
+    if let Some(pos) = name.rfind('.') {
+        name[pos..].to_string()
+    } else {
+        "no ext".to_string()
+    }
+}
+
+fn human_size(bytes: u64) -> String {
```
Comment (F0rty-Tw0) [inline src/cmds/system/ls_win.rs]: `human_size` here is identical to `ls.rs:106-114`. `get_extension` (L141 of this file) duplicates the inline ext-extraction at `ls.rs:183-187`. The output-synthesis logic in `synthesize_output` (L70) reimplements the sort, extension aggregation, and 5-item summary cap from `compact_ls` (`ls.rs:194-230`). Future changes to one path will silently drift from the other.

The design win this PR is positioned to earn: extract a shared formatter that both Linux/Mac (parsed `ls -la`) and Windows (native `fs::read_dir`) feed into. Suggested shape:

```rust
// src/cmds/system/ls_format.rs (new)
pub struct LsRecord { /* moved here */ }
pub fn human_size(bytes: u64) -> String { /* moved here */ }
pub fn get_extension(name: &str) -> String { /* moved here */ }
pub fn format_records(dirs: &[&LsRecord], files: &[&LsRecord]) -> (String, String) { /* shared */ }
```

Then `compact_ls` parses `ls -la` into `Vec<LsRecord>` and calls `format_records`; `ls_win::run_native` calls `fetch_entries` then `format_records`. One formatter, one place to add features (color, glob filtering, etc.) later.

The `LsRecord` struct as you have defined it is already the right data contract; this refactor is mostly about who owns it.

File: src/cmds/system/ls.rs [hunk 3]
```diff
@@ -34,6 +34,13 @@ pub fn run(args: &[String], verbose: u8) -> Result<i32> {
         .map(|s| s.as_str())
         .collect();
 
+    #[cfg(windows)]
+    {
+        if !tool_exists("ls") {
+            return super::ls_win::run_native(&paths, show_all);
```
Comment (F0rty-Tw0) [inline src/cmds/system/ls.rs]: `flags` (collected at L26-L30) is dropped here. `run_native(&paths, show_all)` takes only `paths` and `show_all`, so a user who runs `rtk ls -t` on Windows gets the default sort with no warning. Same for `-r`, `-S`, `-h`, etc.

Two reasonable options:

- **Pass-through with a one-time stderr warning** when unsupported flags accompany the native path, so the user knows their flags did not take effect: `eprintln!("rtk ls: native Windows path ignores flags: {flags:?}")`.
- **Plumb selected flags through** to `run_native(paths, show_all, sort_mode, reverse)` and implement the most common ones (`-t`, `-r`) natively.

Either path is fine. The current silent-drop is what should be fixed.

File: src/cmds/system/ls_win.rs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        }
+        summary.push(')');
+    }
+    summary.push('\n');
+
+    (entries, summary)
+}
+
+/// Entry point called by ls::run on Windows.
+pub fn run_native(paths: &[&str], show_all: bool) -> Result<i32> {
```
Comment (F0rty-Tw0) [inline src/cmds/system/ls_win.rs]: The non-Windows path at `ls.rs:70-102` uses `runner::run_filtered(...)` with `RunOptions::stdout_only().early_exit_on_failure().no_trailing_newline()`, which gives consistent verbose-mode logging (the `Chars: X to Y (Z% reduction)` style line at `ls.rs:85-95`) and a uniform tracking call. `run_native` here calls `print!` directly, hand-starts a `TimedExecution`, and passes `""` for raw input at L134, so `rtk gain --history` will show 0% savings forever for Windows-native ls.

Suggested fixes (any one is enough):

- **Reuse `runner::run_filtered`** with a synthetic raw, passing the entry-list as if it came from `ls -la`, so the runner's verbose and tracking pipeline still applies.
- **Factor the runner's tracking call** into a small helper (`tracking::record(&native_cmd, &cooked, &out)`) and call it from both paths.
- **At minimum**, build a representative `raw` string for the tracking call so the savings metric is not pinned at 0%.

File: src/cmds/system/ls_win.rs [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        entries.push_str(&format!("{}/\n", d.name));
+    }
+    for f in &files {
+        entries.push_str(&format!("{}  {}\n", f.name, human_size(f.size)));
+    }
+
+    let mut summary = format!("\nSummary: {} files, {} dirs", files.len(), dirs.len());
+    if !by_ext.is_empty() {
+        let mut ext_counts: Vec<_> = by_ext.iter().collect();
+        ext_counts.sort_by(|a, b| b.1.cmp(a.1));
```
Comment (F0rty-Tw0) [inline src/cmds/system/ls_win.rs]: Two small correctness items at the synthesis layer:

**Tied-extension ordering is non-deterministic.** The visible Linux/Windows summary divergence in the PR description (`1 .gitignore, 1 .toml, 1 .lock` vs `1 .lock, 1 no ext, 1 .sh`) is this sort: items with equal counts come out in HashMap iteration order. Fix: `ext_counts.sort_by(|a, b| b.1.cmp(a.1).then(a.0.cmp(b.0)))` to break ties by extension name. Pre-existing in `compact_ls` too, so the unified formatter from the F2 thread fixes both at once.

**Symlinks lose their target.** `compact_ls` parses `link -> target` from `ls -la` output (see `ls.rs:182` and the `test_compact_symlinks` test at `ls.rs:326-331`). The native path uses `metadata.is_dir()` only, so a symlink to a dir renders as `link/` and a symlink to a file as `link size`, target name dropped. `metadata.file_type().is_symlink()` plus `std::fs::read_link()` would let `synthesize_output` carry the same `link -> target` rendering. Worth at least a test that documents current behavior.

**Comment by seansie0830** [comment]
Thanks for the insightful review, @F0rty-Tw0! I’ve acknowledged all the P0/P1 issues. To keep the momentum going while I’m a bit tied up, I’ve drafted a Fixing Roadmap below.

I have enabled maintainer edits, so other contributors are more than welcome to push commits directly to this branch. I’d be happy to collaborate on this and will also jump back in to help with the implementation as soon as I have a breather!

🛠 Fixing Roadmap & Checklist
- [ ] Decouple Logic (Backend vs. Frontend)

Extract a shared ls_format.rs (or similar) to act as the "Frontend" formatter.

Refactor both the Unix parser and the Windows native fs logic to output a standardized Vec<LsRecord>.

Unify human_size, sorting logic (with tie-breaking), and extension summary in one place.

- [ ] Enhance Native Fallback Transparency

Add a warning message (stderr) when native fallback is triggered if the user provided unsupported flags (e.g., -t, -r).

- [ ] Integrate with Runner & Tracking

Ensure run_native uses the internal runner or tracking mechanism so that token-savings and execution time are correctly recorded in rtk gain --history.

- [ ] Support Symlinks in Native Path

Update fetch_entries to handle symlink detection and rendering (link -> target) to match the Unix output.

- [ ] Add Test Coverage

Add Snapshot tests for the new shared formatter using a mock Vec<LsRecord>.

Add unit tests for fetch_entries using tempfile to verify filtering of NOISE_DIRS.

File: src/cmds/system/ls_win.rs (hunk 1 above)
Comment (F0rty-Tw0) [reply src/cmds/system/ls_win.rs]: Partially resolved. `synthesize_output` is now exercised through the existing `ls.rs` tests since it lives in the shared `ls_format.rs` — that's good. Still outstanding:

- `fetch_entries` has no test (the Windows-specific filesystem traversal, NOISE_DIRS filter, dir/file classification path is unverified).
- No explicit token-savings assertion (≥60%) anywhere for the native path.

A `tempfile::tempdir()`-backed test for `fetch_entries` and a savings test in `ls_format.rs` would close this out.

File: src/cmds/system/ls_win.rs (hunk 2 above)
Comment (F0rty-Tw0) [reply src/cmds/system/ls_win.rs]: Resolved. `ls_format.rs` owns `LsRecord`, `human_size`, and `synthesize_output`; `ls_unix.rs` owns `parse_ls_line`; `ls_win.rs` imports them. This is exactly the shared-formatter shape I suggested — nice work.

File: src/cmds/system/ls.rs (hunk 3 above)
Comment (F0rty-Tw0) [reply src/cmds/system/ls.rs]: Partially resolved. The unconditional warning in `run_native` is better than silent drop, but it fires on every Windows-native run including `rtk ls` with no flags, and never echoes *which* flags were dropped. The message ("some flag may not work as expected") is technically incorrect on a flagless invocation.

Suggested tighten-up: only print when `!flags.is_empty()`, and include the flags in the message — e.g. `eprintln!("rtk ls: native Windows path ignores flags: {:?}", flags)`. Requires plumbing `flags` into `run_native`.

File: src/cmds/system/ls_win.rs (hunk 4 above)
Comment (F0rty-Tw0) [reply src/cmds/system/ls_win.rs]: Not addressed. `run_native` still calls `timer.track(..., "", &output)` at line 82 — empty raw string means `rtk gain --history` will report 0% savings forever for native Windows `ls`, which is the exact behavior I flagged.

Minimum fix: build a representative raw baseline (e.g. synthesize a fake `ls -la`-style line per record before formatting) so the savings metric is meaningful. The dedup refactor actually makes this easier now — both paths can share the same baseline-synthesis helper.

File: src/cmds/system/ls_win.rs (hunk 5 above)
Comment (F0rty-Tw0) [reply src/cmds/system/ls_win.rs]: Tied-extension ordering: resolved. `ls_format.rs:51` adds the `.then_with(|| a.0.cmp(b.0))` tiebreak, which also fixes the same pre-existing bug on the Unix path. Good catch making it the shared fix.

Symlinks: not addressed. `fetch_entries` still classifies purely on `metadata.is_dir()`, so a symlink-to-dir renders as `link/` with the target dropped. `metadata.file_type().is_symlink()` + `std::fs::read_link()` would let `synthesize_output` carry the same `link -> target` rendering as the Unix path. Even just a regression test documenting the current behavior would help.

**Comment by seansie0830** [comment]
### update note for [d900c04](https://github.com/rtk-ai/rtk/pull/1366/commits/d900c04cb2222f1dc577ab632127c53f23ff8f40) commit 
### 📝 PR Update Note
Note: E2E tests are planned for a follow-up task.

### 🛠 Architecture: Decoupling Logic
Refactored the ls module to improve maintainability across platforms:

Unified Data Structure: Moved the core data contract to ls.rs as the shared schema for both Windows and Unix.

Frontend Logic: Consolidated all formatting logic into ls_format.

Platform-specific Backends: Isolated logic into ls_win and ls_unix respectively.

### 🚀 Native Fallback & MVP Features
Implemented an MVP for native fallback when the ls tool is missing on Windows:

Supported Flags: Added support for common flags: -t and -r.

Transparency: Added warn_unsupported_flags in ls_win:126 to inform users of ignored flags.

### ⏱ Integrate with Runner & Tracking
To avoid process overhead, I switched from runner::run_filtered (a new process is required)  to a more direct integration using timer.track. This allows for high-performance execution while maintaining telemetry.

### 🔗 Symlink Support & Testing
Symlink Handling: Implemented and verified symlink support on Windows.

Testing: Added comprehensive tests in ls_win.rs, specifically test_fetch_entries_symlinks using the tempfile crate to simulate file systems.

Coverage: Added Windows-exclusive test suites at the end of ls_win.rs.
> Snapshot testing is deferred until the logic is fully stabilized.

**Comment by seansie0830** [comment]
Hi everyone! 👋

I've pushed the latest updates to **PR #1366** (Native Windows `ls` fallback & SRP refactor).

* **PR:** https://github.com/rtk-ai/rtk/pull/1366
* **Latest commit:** https://github.com/seansie0830/rtk/commit/19235ae20ec42cd3afb765d7d4c2ec8b19143a28

### Highlights

* ✅ Resolved all review feedback from **@F0rty-Tw0**:

  * Symlink handling
  * `-t` / `-r` flags
  * Deterministic sorting
  * Token savings of **≥60%**
* ✅ Fixed standalone `-h` flag parsing.
* ✅ Verified **1,460/1,460** unit tests passing.
* ✅ Added an automated **9-part E2E** suite (`e2e_ls_native.ps1`) — **100% green** 🟢.

Whenever you have some time, I'd really appreciate a review. Thanks! 🙌
